### PR TITLE
implement the separated incrementally loading listview

### PR DESCRIPTION
**Pull request by avdosev**
Hi, I implemented a separated incrementally loading listview. I needed it in my project and decided to share it.

### CODE REVIEW

**Review by MaikuB — CHANGES_REQUESTED** [review]
Could you add API docs for this and also update the example app to demonstrate its usage? This could involve changing the app to start with a landing page where users pick between going to a page that displays the standard or separated list view

File: lib/src/incrementally_loading_listview.dart [hunk 1]
```diff
@@ -68,6 +68,44 @@ class IncrementallyLoadingListView extends StatefulWidget {
...
+      @required IndexedWidgetBuilder itemBuilder,
+      @required IndexedWidgetBuilder separatorBuilder,
+      @required ItemCount itemCount,
+      this.addAutomaticKeepAlives = true,
+      this.addRepaintBoundaries = true,
+      this.cacheExtent,
+      this.onLoadMore,
+      this.onLoadMoreFinished})
+      : this.itemCount = (() {
+          final count = itemCount();
```
Comment (MaikuB) [inline lib/src/incrementally_loading_listview.dart]: .Sorry noticed something else where what the user is passing for `itemCount` is getting changed. I get why it is done but this is a public property and is something the user specified so should be kept as is. Same thing for `itemBuilder` That the number of items is being increased is something that is better handled internally. By that mean, to see if all of this logic could be moved to be around where the widget gets built and calls `ListView.builder(...)`. IMO, it's better to do it this way as some of the logic here is related to what should get rendered so having it in the `build` method is more appropriate

**Comment by avdosev** [comment]
Ok, I corrected the code based on your comments. Is that what you mean?

**Review by MaikuB — CHANGES_REQUESTED** [review]
There are some changes here that weren't expected/inline with what was I referring to. Moving part of the logic to be part of the `build` method was done but didn't expect to have had the `separated` named constructor removed. If you could add that back then that would be great as it is more inline with what the Flutter SDK provides (i.e. helps to follow the convention even though there's no `builder` named constructor here) and the use of a named constructor can help improve readability. As it is right now, the optional `separatorBuilder`  makes the feature less discoverable whils using the `separated` named constructor makes the it required parameter.  So to reiterate/summarise the old constructor in the package remains as is but a `IncrementallyLoadingListView.separated` named constructor like you had before would exist.

Changing this would mean the changes you have in the readme gets reverted but can mentioned that the ability to provide a separator.

With the changes to the example app, I know it's a simple app but the use of an icon doesn't much of an explanation on what it's for. Tooltips are only shown when users do a long-press and it's rare for users to do that on a button. Better to be explicit about what it does via text than them hitting a button to then figure out what it does. Hence my comment on changing the example app to have a landing page with buttons where users can go to different pages, one for the original widget and one for the separated one you've added in this PR. By doing it in this manner, the buttons themselves could have text

**Comment by avdosev** [comment]
Ok. At the moment is that what you mean?

**Comment by MaikuB** [comment]
Forgot to come back to this but not quite. Will see if I can address this myself after #21 as null safety support seems more important

**Comment by avdosev** [comment]
It's easier and faster to write new a package based on a fork instead of wait you.

**Comment by MaikuB** [comment]
Removed the separated named constructor as it looks like having that forced the original constructor to require taking in the new property as well, which adds redundancy. The thought of having the named constructor was to have similar options to the `ListView` widget.

As for your comment on being faster to write a new package etc, perhaps so. However, a reminder that open-source software is provided as is and a lot of people doing open-sourced work do so in their free time and may have other priorities in their personal life. If you intend to engage with more of the community in the future then I'd suggest you reconsider making snarky comments like that as the community would be likely less willing engage with you.

**Pull request merged**